File: src/test/scala/com/xebia/functional/prompt/ConfigSpec.scala
package com.xebia.functional.prompt

import cats.effect.*

import com.xebia.functional.prompt.models.*
import munit.CatsEffectSuite

class ConfigSpec extends CatsEffectSuite:

  test("make should return a valid Config if the template and input variables are valid") {

    val template = "Hello {name}, you are {age} years old."
    val inputVariables = List("name", "age")
    val configIO = Config.make[IO](template, inputVariables)

    val config = configIO.unsafeRunSync()
    assertEquals(config.inputVariables, inputVariables)
    assertEquals(config.template, template)
    assertEquals(config.templateFormat, TemplateFormat.FString)
  }

  test("make should fail with a InvalidTemplateError if the template has missing arguments") {
    val template = "Hello {name}, you are {age} years old."
    val inputVariables = List("name")
    val configIO = Config.make[IO](template, inputVariables)

    interceptMessageIO[InvalidTemplateError]("Template 'Hello {name}, you are {age} years old.' has missing arguments: {age}")(configIO)
  }

  test("make should fail with a InvalidTemplateError if the template has unused arguments") {
    val template = "Hello {name}, you are {age} years old."
    val inputVariables = List("name", "age", "city")
    val configIO = Config.make[IO](template, inputVariables)

    interceptMessageIO[InvalidTemplateError]("Template 'Hello {name}, you are {age} years old.' has unused arguments: {city}")(configIO)
  }

  test("make should fail with a InvalidTemplateError if there are duplicate input variables") {
    val template = "Hello {name}, you are {name} years old."
    val inputVariables = List("name")
    val configIO = Config.make[IO](template, inputVariables)

    interceptMessageIO[InvalidTemplateError]("Template 'Hello {name}, you are {name} years old.' has duplicate arguments: {name}")(configIO)
  }

  test("make should fail with a combination of InvalidTemplateErrors if there are multiple things wrong") {
    val template = "Hello {name}, you are {name} years old."
    val inputVariables = List("name", "age")
    val configIO = Config.make[IO](template, inputVariables)

    interceptMessageIO[InvalidTemplateError](
      "Template 'Hello {name}, you are {name} years old.' has unused arguments: {age}; " +
        "Template 'Hello {name}, you are {name} years old.' has duplicate arguments: {name}"
    )(configIO)
  }

File: src/commonTest/kotlin/com/xebia/functional/prompt/ConfigSpec.kt
package com.xebia.functional.prompt

import arrow.core.raise.either
import io.kotest.assertions.arrow.core.shouldBeLeft
import io.kotest.assertions.arrow.core.shouldBeRight
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe

class ConfigSpec : StringSpec({

  "should return a valid Config if the template and input variables are valid" {
    val template = "Hello {name}, you are {age} years old."
    val variables = listOf("name", "age")

    val config = either { Config(template, variables) }.shouldBeRight()

    config.inputVariables shouldBe variables
    config.template shouldBe template
    config.templateFormat shouldBe TemplateFormat.FString
  }

  "should fail with a InvalidTemplateError if the template has missing arguments" {
    val template = "Hello {name}, you are {age} years old."
    val variables = listOf("name")

    either {
      Config(template, variables)
    } shouldBeLeft InvalidTemplate("Template 'Hello {name}, you are {age} years old.' has missing arguments: {age}")
  }

  "should fail with a InvalidTemplateError if the template has unused arguments" {
    val template = "Hello {name}, you are {age} years old."
    val variables = listOf("name", "age", "unused")

    either {
      Config(template, variables)
    } shouldBeLeft InvalidTemplate("Template 'Hello {name}, you are {age} years old.' has unused arguments: {unused}")
  }

  "should fail with a InvalidTemplateError if there are duplicate input variables" {
    val template = "Hello {name}, you are {name} years old."
    val variables = listOf("name")

    either {
      Config(template, variables)
    } shouldBeLeft InvalidTemplate("Template 'Hello {name}, you are {name} years old.' has duplicate arguments: {name}")
  }

  "should fail with a combination of InvalidTemplateErrors if there are multiple things wrong" {
    val template = "Hello {name}, you are {name} years old."
    val variables = listOf("name", "age")
    val unused = "Template 'Hello {name}, you are {name} years old.' has unused arguments: {age}"
    val duplicated = "Template 'Hello {name}, you are {name} years old.' has duplicate arguments: {name}"

    either {
      Config(template, variables)
    } shouldBeLeft InvalidTemplate("$unused, $duplicated")
  }
})
